service/src/main/kotlin/fi/espoo/evaka/attendance/StaffAttendanceQueries.kt

```diff
@@ -406,6 +407,21 @@ WHERE employee_id = :employeeId AND tstzrange(start_time, end_time) && tstzrange
         .mapTo<PlannedStaffAttendance>()
         .toList()
 
+fun Database.Read.getPlannedStaffAttendanceForDays(
+    employeeIds: List<EmployeeId>,
+    days: List<LocalDate>
+): List<PlannedStaffAttendanceWithEmployeeId> =
+    createQuery(
+        """
+SELECT start_time AS start, end_time AS end, type, employee_id FROM staff_attendance_plan
+WHERE employee_id = ANY(:employeeIds) AND (date(start_time) = ANY(:days) OR date(end_time) = ANY(:days))
```

Comment:
This query can't use an index for the time part (= database fetches all plans for employees regardless of time, and filters one-by-one), but there's a fix:

Take a FiniteDateRange instead of a list of arbitrary dates, convert it to two HelsinkiDateTimes, and construct a tstzrange:

```kotlin
.bind("startTime", HelsinkiDateTime.of(range.start, LocalTime.MIDNIGHT))
.bind("endTime", HelsinkiDateTime.of(range.end.plusDays(1), LocalTime.MIDNIGHT))
```

and

```sql
WHERE employee_id = ANY(:employeeIds) AND (tstzrange(start_time, end_time) && tstzrange(:start_time, :end_time))
```

PS. You can also accept `Collection<EmployeeId>` so you can pass `attendancesByEmployee.keys` directly in the call site without the extra `.toList()` call

Reply:
This makes more sense than my original query too as start_time and end_time forms a range, so if the plan was over 2 days long the query would fail to find the plans in the middle, whereas by using tstzrange the overlap is checked properly. Thanks!